File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.ts
function shipWithinDays(weights: number[], days: number): number {
    let l = Math.max(...weights);
    let r = weights.reduce((n, accum) => n + accum, 0);
    let ans = Infinity;

    const isValid = (capacity: number): boolean => {
        let curSum = 0;
        let curDays = 1;
        for (let i = 0; i < weights.length; i++) {
            if (curSum + weights[i] > capacity) {
                curDays++;
                curSum = weights[i];
            }
            else curSum += weights[i];
        }
        return curDays <= days;
    }

    while (l <= r) {
        let m = Math.floor((l + r) / 2);
        if (isValid(m)) {
            ans = Math.min(ans, m);
            r = m - 1;
        }
        else l = m + 1;
    }
    return ans;
};

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int l = 0;
        int r = 0;
        int ans = 0;
        for(int weight : weights) {
            l = Math.max(l , weight);
            r += weight;
        }
        while(l <= r) {
            int m = l + (r - l) / 2;
            int requiredDays = findDaysGivenWeight(weights , m);
            if(requiredDays > days) l = m + 1;
            else {
                ans = m;
                r = m - 1;
            }
        }
        return ans;
    }
    private int findDaysGivenWeight(int[] weights , int minW) {
        int days = 0;
        int curWeight = 0;
        for(int i = 0; i < weights.length; i ++) {
            if(curWeight + weights[i] <= minW) curWeight += weights[i];
            else {
                curWeight = weights[i];
                days ++;
            }
        }
        days ++;
        return days;
    }
}